source/MRViewer/MRSurfaceContoursWidget.cpp

```diff
@@ -0,0 +1,407 @@
+std::string ChangePointActionPickerPoint::name() const
+{
+    return "Move Point";
+}
+void ChangePointActionPickerPoint::action( Type )
+{
+    if ( !widget_.isPickerActive_ )
+        return;
+
+    widget_.pickedPoints_[obj_][index_]->updateCurrentPosition( point_ );
+    widget_.activeIndex = index_;
+    widget_.activeObject = obj_;
+    widget_.onPointMoveFinish_( obj_ );
+}
+
+size_t ChangePointActionPickerPoint::heapBytes() const
+{
+    return 0; //this undo action will be deleted in widget disable
+}
+
+void SurfaceContoursWidget::enable( bool isEnaled )
+{
+    isPickerActive_ = isEnaled;
+    if ( !isPickerActive_ )
+        pickedPoints_.clear();
+}
+
+std::shared_ptr<SurfacePointWidget> SurfaceContoursWidget::createPickWidget_( const std::shared_ptr<MR::ObjectMeshHolder>& obj, const MeshTriPoint& pt )
+{
+    auto newPoint = std::make_shared<SurfacePointWidget>();
+    newPoint->setAutoHover( false );
+    newPoint->setParameters( params.surfacePointParams );
+    auto objMesh = std::dynamic_pointer_cast< MR::ObjectMesh > ( obj );
+    if ( objMesh )
+        newPoint->create( objMesh, pt );
+    else
+    {
+        assert( 1 > 2 ); // due to surfacePointWidget unable to work with non ObjectMesh objects.
+        return {};
+    }
+    std::weak_ptr<SurfacePointWidget> curentPoint = newPoint;
+    newPoint->setStartMoveCallback( [this, obj, curentPoint] ( const MeshTriPoint& point )
+    {
+        const bool closedPath = isClosedCountour( obj );
+
+        if ( closedPath && curentPoint.lock() == pickedPoints_[obj][0] )
+        {
+            if ( params.writeHistory )
+            {
+                SCOPED_HISTORY( "Change Point" );
+                AppendHistory<ChangePointActionPickerPoint>( *this, obj, point, activeIndex );
+                AppendHistory<ChangePointActionPickerPoint>( *this, obj, point, int( pickedPoints_[obj].size() ) - 1 );
+            }
+            moveClosedPoint_ = true;
+        }
+        else if ( !closedPath || curentPoint.lock() != pickedPoints_[obj].back() )
+        {
+            if ( params.writeHistory )
+                AppendHistory<ChangePointActionPickerPoint>( *this, obj, point, activeIndex );
+        }
+        activeChange_ = true;
+        onPointMove_( obj );
+
+    } );
+    newPoint->setEndMoveCallback( [this, obj, curentPoint] ( const MeshTriPoint& point )
+    {
+        if ( moveClosedPoint_ && curentPoint.lock() == pickedPoints_[obj][0] )
+        {
+            pickedPoints_[obj].back()->updateCurrentPosition( point );
+            moveClosedPoint_ = false;
+        }
+        activeChange_ = false;
+        onPointMoveFinish_( obj );
+    } );
+
+    return newPoint;
+}
+
+bool SurfaceContoursWidget::isClosedCountour( const std::shared_ptr<ObjectMeshHolder>& obj )
+{
+    return ( ( pickedPoints_.count( obj ) > 0 )
+        && ( pickedPoints_[obj].size() > 1 )
+        && ( pickedPoints_[obj][0]->getCurrentPosition() == pickedPoints_[obj].back()->getCurrentPosition() ) );
```

Comment:
several find operation in map, better to do find only once
```c++
auto pointsIt = pickedPoints_.find( obj );
if ( pointsIt == pickedPoints_.end() )
    return false;
return pointsIt->second.size() > 1 && pointsIt->second[0]->getCurrentPosition() == pointsIt->second.back()->getCurrentPosition();

Reply:
update here and in other plases.